include/Transforms/Functional.h

```diff
@@ -0,0 +1,487 @@
+  /// Return the contained value.
+  static type unpack(FailureOr<type> &&value) { return std::move(*value); }
+};
+} // end namespace detail
+
+/// Given a scope, apply a pattern once per operation in the scope, saving the
+/// result of each match. The result list is empty if the pattern failed to
+/// match at all.
+template <typename PatternT, typename... Args,
+          bool = PatternConcept<PatternT>::verify()>
+auto applyEach(Operation *scope, PatternT &&pattern, Args &&...args) {
+  using Traits = llvm::function_traits<std::decay_t<PatternT>>;
+  using Unpack = detail::UnpackFailureOr<typename Traits::result_t>;
+
+  assert(scope->hasTrait<OpTrait::IsIsolatedFromAbove>() &&
+         "scope is not isolated from above");
+
+  detail::SimpleRewriter rewriter(scope->getContext());
+  // A list of all the results.
+  SmallVector<typename Unpack::type> results;
+  scope->walk([pattern = std::forward<PatternT>(pattern), &rewriter, &results,
+               &args...](Operation *op) {
+    rewriter.setInsertionPoint(op);
+    auto result = pattern(op, rewriter, std::forward<Args>(args)...);
+    // If the pattern applied, unpack the result and store it.
+    if (succeeded(result))
+      results.push_back(Unpack::unpack(std::move(result)));
+  });
+  return results;
+}
+
+/// Apply a pattern directly on an operation, for each operation in a list.
+template <typename ListT, typename PatternT, typename... Args,
+          bool = PatternConcept<PatternT>::verify()>
+auto applyForEach(ListT &&list, PatternT &&pattern, Args &&...args) {
+  using Traits = llvm::function_traits<std::decay_t<PatternT>>;
+  using Unpack = detail::UnpackFailureOr<typename Traits::result_t>;
+
+  // A list of all the results.
+  SmallVector<typename Unpack::type> results;
+  for (auto op : list) {
+    auto result = applyAt(op, std::forward<PatternT>(pattern),
+                          std::forward<Args>(args)...);
+    // The pattern applied, unpack the result and store it.
+    if (succeeded(result))
+      results.push_back(Unpack::unpack(std::move(result)));
+  }
+  return results;
+}
+
+/// Find the nearest proper ancestor that is isolated from above.
+Operation *findNearestScope(Operation *op);
+
+/// Clone the regions of an operation.
+SmallVector<std::unique_ptr<Region>> cloneRegionsOf(Operation *op);
+/// Replace the regions of an operation with the given regions.
+void takeRegions(Operation *op, ArrayRef<std::unique_ptr<Region>> regions);
+
+namespace detail {
+template <typename OpT, typename SequenceT, typename... Args,
+          bool = PatternConcept<SequenceT>::verify()>
+auto applySequence(PatternRewriter &rewriter, Operation *scope, OpT op,
+                   SequenceT &&sequence, Args &&...args) {
+  // Save a copy of the scope.
+  SmallVector<std::unique_ptr<Region>> backup = cloneRegionsOf(scope);
+
+  rewriter.setInsertionPoint(op);
+  auto result = std::forward<SequenceT>(sequence)(op, rewriter,
```

Comment:
Could you add a comment here to say what this is forwarded to? I am not clear which of the other functions can get called as the result of that.

I am asking because this relates to your generic findNearestScope but I am wondering whether some paths may bottom out on the wrap/unwrap whose impl. is very specific to ScopeOp.
Outlining generic isolated form above ops can get quite painful (i.e. finding and reconciling terminators)